### PR TITLE
Fracture user's GAS UTXOs

### PR DESCRIPTION
**Pull request by jameswhollister**
## Ticket
*924b2140a4

## Changes
* Fracture user's GAS UTXOs
* Cache system asset balances in memory
* User known unspent gas UTXOs
* Prompt user, asking whether they want to fracture their GAS

## Screenshots
image

### CODE REVIEW

File: src/renderer/services/neo.js [hunk 1]
```diff
@@ -908,6 +905,223 @@ export default {
...
+
+        if (!forAddress) {
+          const currentWallet = wallets.getCurrentWallet();
+          forAddress = currentWallet.address;
+        }
+
+        if (_.has(addressBalances, forAddress)) {
+          const existingBalance = _.get(addressBalances, forAddress);
+          if (existingBalance && existingBalance.pulled
+            && existingBalance.isExpired !== true
```
[the remark this answers is not in the document]
Comment (jameswhollister) [reply src/renderer/services/neo.js]: This concept isn't needed anymore, changed my method to reset the cache if a new TX comes in.

File: src/renderer/services/neo.js [hunk 2]
```diff
@@ -908,6 +905,223 @@ export default {
...
+            && moment().utc().diff(existingBalance.pulled, 'milliseconds') < timeouts.BALANCE_PERSIST_FOR) {
```
[the remark this answers is not in the document]
Comment (jameswhollister) [reply src/renderer/services/neo.js]: Correct, this is more of a safeguard

File: src/renderer/components/modals/FractureGasModal.vue [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import ModalWrapper from './ModalWrapper';
+
+export default {
+  components: {
+    ModalWrapper,
+  },
+
+  computed: {
+    shouldDisableFractureButton() {
+      return this.sending;
```
Comment (crishellco) [inline src/renderer/components/modals/FractureGasModal.vue]: This 👍

File: src/renderer/services/dex.js [hunk 4]
```diff
@@ -1299,17 +1299,21 @@ export default {
         const token = assets.getNetworkAsset(assetId);
 
         api.fillKeys(config)
-          .then(() => {
-            return api.getBalanceFrom({
-              net: currentNetwork.net,
-              url: currentNetwork.rpc,
-              address: wallet.getAddressFromScriptHash(assets.DEX_SCRIPT_HASH),
-            }, api.neoscan);
+          .then((c) => {
```
Comment (crishellco) [inline src/renderer/services/dex.js]: Variable name 😏

File: src/renderer/services/dex.js [hunk 5]
```diff
@@ -1299,17 +1299,21 @@ export default {
...
+            return new Promise((resolveBalance) => {
+              neo.fetchSystemAssetBalance(currentWallet.address)
+                .then((balance) => {
+                  c.balance = balance;
+                  resolveBalance(c);
+                })
+                .catch((e) => {
```
Comment (crishellco) [inline src/renderer/services/dex.js]: Reject message 😏

File: src/renderer/services/dex.js [hunk 6]
```diff
@@ -2006,7 +2009,16 @@ export default {
...
             }
 
-            return api.getBalanceFrom(c, api.neoscan);
+            return new Promise((resolveBalance) => {
+              neo.fetchSystemAssetBalance(currentWallet.address, config.intents)
+                .then((balance) => {
+                  c.balance = balance;
+                  resolveBalance(c);
+                })
+                .catch((e) => {
```
Comment (crishellco) [inline src/renderer/services/dex.js]: Reject message 😏

File: src/renderer/services/neo.js [hunk 7]
```diff
@@ -931,6 +928,264 @@ export default {
...
+            currentOutputsAboveFee: outputsAboveFee,
+            recommendedUTXOs,
+            fee: currentNetwork.fee,
+          });
+
+          lastGasFractureNotification = new Date();
+        }
+      })
+      .catch((e) => {
+        alerts.exception(e);
```
Comment (crishellco) [inline src/renderer/services/neo.js]: Alert message 😏

File: src/renderer/services/neo.js [hunk 8]
```diff
@@ -931,6 +928,264 @@ export default {
...
+                const dump = {
+                  net: config.net,
+                  address: config.address,
+                  intents: config.intents,
+                  balance: config.balance,
+                  script: config.script,
+                  gas: config.gas,
+                  tx: config.tx,
+                };
+                console.log(dump);
```
Comment (crishellco) [inline src/renderer/services/neo.js]: Do we need this `console.log`?

**Comment by rlucas6130** [comment]
@jameswhollister 

Before I click 'Yes, Fracture GAS':
image

After:
image

**Comment by jameswhollister** [comment]
@rlucas6130 the two issues you described should be resolved with the latest

**Comment by rlucas6130** [comment]
Looks like there is a console.log in there we might want to remove.

**Comment by jsolman** [comment]
Mainly just waiting for an adjustment to break GAS into nearly equal size pieces before merge

**Pull request merged**